adapters/repos/db/lsmkv/snapshot.go

```diff
@@ -0,0 +1,84 @@
+package lsmkv
+
+import (
+	"context"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// PauseCompaction waits for all ongoing compactions to finish,
+// then makes sure that no new compaction can be started.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// compactions are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) PauseCompaction(ctx context.Context) error {
+	compactionHalted := make(chan struct{})
+
+	go func() {
+		b.disk.stopCompactionCycle <- struct{}{}
+		compactionHalted <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return errors.Wrap(ctx.Err(), "long-running compaction in progress")
+	case <-compactionHalted:
+		return nil
+	}
+}
+
+// FlushMemtable flushes any active memtable and returns only once the memtable
+// has been fully flushed and a stable state on disk has been reached.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// flushes are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) FlushMemtable(ctx context.Context) error {
+	b.stopFlushCycle <- struct{}{}
+	defer b.initFlushCycle()
+
+	if b.active == nil && b.flushing == nil {
+		return nil
+	}
+
+	return b.FlushAndSwitch()
+}
+
+// ListFiles lists all files that currently exist in the Bucket. The files are only
+// in a stable state if the memtable is empty, and if compactions are paused. If one
+// of those conditions is not given, it errors
```

Comment:
I wonder if stable state check (paused compaction, flushed memtables) should be done inside the method or on a higher level. At the moment it seems to be just listing files

Reply:
Yes, from my discussions with @etiennedi coordination such you have described will be implemented on the shard/index level.